### PR TITLE
Split dlr_internal.h and dlr.c into multiple files

### PR DESCRIPTION
**Pull request by apivovarov**
`dlr_internal.h` and `dlr.c` were split into multiple files.
Now DLR has the following header and cc files
```
dlr.h # Public C API
dlr_common.h   # common stuff for DLR Developers (DLRModel Abstract class, etc)
dlr_tvm.h   # TVMModel class
dlr_treelite.h   # TreeliteModel class

dlr.cc # Contains only public C API implementation (matches to `dlr.h`)
dlr_common.cc   # GetBasename, listdir, get_backend impl
dlr_tvm.cc
dlr_treelite.cc
```

`dlr.cc` has `CreateDLRModel` functions which decides what `DLRModel` implementation to create

`dlr_common.h` uses some headers from `3rdparty`:
`3rdparty/tvm/src/runtime_base.h`  (for `DLContext` and `API_BEGIN()`, `API_END()`)
`dmlc/logging.h` (for `LOG(FATAL)`)

`dlr_common.cc` uses:
`dmlc/filesystem.h` - for `dmlc::io::FileSystem` used by `dlr::listdir` impl

There were no changes in public C API defined in dlr.h.

New `libdlr.so` was tested using [neo-ai-dlr-test](https://github.com/apivovarov/neo-ai-dlr-test) project. It uses DLR API to run compiled Resnet18v2 model.

### CODE REVIEW

File: src/dlr_common.cc [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#else
+  char* path_ = strdup(path.c_str());
+  char* base = basename(path_);
+  std::string ret(base);
+  free(path_);
+  return ret;
+#endif
+}
+
+void dlr::listdir(const std::string& dirname, std::vector<std::string> &paths) {
```
Comment (zhiics) [inline src/dlr_common.cc]: Let's use CamelCase for naming to keep consistency.

File: src/dlr_common.cc [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  std::vector<dmlc::io::FileInfo> file_list;
+  fs->ListDirectory(uri, &file_list);
+  for (dmlc::io::FileInfo info : file_list) {
+    if (info.type != dmlc::io::FileType::kDirectory) {
+      paths.push_back(info.path.name);
+    }
+  }
+}
+
+DLRBackend dlr::get_backend(const std::string &dirname) {
```
Comment (zhiics) [inline src/dlr_common.cc]: ditto

File: src/dlr_treelite.cc [hunk 3]
```diff
@@ -0,0 +1,184 @@
+#include "dlr_treelite.h"
+#include <cmath>
+#include <cstring>
+#include <fstream>
+
+
+using namespace dlr;
+
+// not used (was used in SetupTreeliteModule)
+std::string get_version(const std::string &json_path) {
```
Comment (zhiics) [inline src/dlr_treelite.cc]:
```suggestion
std::string GetVersion(const std::string& json_path) {
```

File: src/dlr_treelite.cc [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+      else if (colon_flag && quote_flag && (c == '"'))
+        return version;
+      else if (colon_flag && quote_flag)
+        version.push_back(c);
+    }
+  }
+  return version;
+}
+
+ModelPath dlr::get_treelite_paths(const std::string &dirname) {
```
Comment (zhiics) [inline src/dlr_treelite.cc]:
```suggestion
ModelPath dlr::GetTreelitePaths(const std::string& dirname) {
```

File: include/dlr_tvm.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+#define DLR_TVM_H_
+
+#include <graph/graph_runtime.h>
+#include "dlr_common.h"
+
+namespace dlr {
+
+/*! /brief Get the paths of the TVM model files.
+ */
+ModelPath get_tvm_paths(const std::string &tar_path);
```
Comment (zhiics) [inline include/dlr_tvm.h]: Let's use the CamelCase to keep consistency.

ModelPath GetTVMPaths(const std::string& tar_path);

File: include/dlr_treelite.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  std::vector<uint32_t> col_ind;
+  std::vector<size_t> row_ptr;
+  size_t num_row;
+  size_t num_col;
+  CSRBatchHandle handle;
+};
+
+/*! /brief Get the paths of the Treelite model files.
+ */
+ModelPath get_treelite_paths(const std::string& dirname);
```
Comment (zhiics) [inline include/dlr_treelite.h]: Let's use CamelCase to keep consistency, `GetTreeLitePath`

File: include/dlr_common.h [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+enum class DLRBackend {
+  kTVM,
+  kTREELITE
+};
+
+/*! /brief Get the backend based on the contents of the model folder.
+ */
+DLRBackend get_backend(const std::string &dirname);
```
Comment (zhiics) [inline include/dlr_common.h]: Let's use CamelCase for function names to keep consistency.

`DLRBackend GetBackend(const std::string& dirname);`

File: include/dlr_common.h [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  std::string params;
+  std::string model_json;
+  std::string ver_json;
+} ModelPath;
+
+void listdir(const std::string& dirname, std::vector<std::string> &paths);
+
+std::string GetBasename(const std::string& path);
+
+inline bool endsWith(const std::string &mainStr, const std::string &toMatch) {
```
Comment (zhiics) [inline include/dlr_common.h]: Let's use CamelCase for function names to keep consistency.

`inline bool EndsWith(const std::string& main_str, const std::string& to_match)`

**Comment by zhiics** [comment]
Mostly looks good to me. Just some stylish comments.

**Comment by apivovarov** [comment]
The goal of this PR is to split existing `dlr.cc` and `dlr_internal.h` files into 7 new files.
In this particular PR I tried to minimize code changes and I only changed `#include` blocks to include what is actually necessary for new .cc/.h files.
Lets do method/function names as well as pointer/reference sing position standardization in a separate PR.
I believe we should  use automatic style/convention checker or code analyzer to do this job instead of doing it manually.

File: src/dlr_common.cc (hunk 1 above)
Comment (apivovarov) [reply src/dlr_common.cc]: fixed

File: src/dlr_common.cc (hunk 2 above)
Comment (apivovarov) [reply src/dlr_common.cc]: fixed

File: src/dlr_treelite.cc (hunk 3 above)
Comment (apivovarov) [reply src/dlr_treelite.cc]: fixed

File: src/dlr_treelite.cc (hunk 4 above)
Comment (apivovarov) [reply src/dlr_treelite.cc]: fixed

File: include/dlr_tvm.h (hunk 5 above)
Comment (apivovarov) [reply include/dlr_tvm.h]: fixed

File: include/dlr_treelite.h (hunk 6 above)
Comment (apivovarov) [reply include/dlr_treelite.h]: fixed

File: include/dlr_common.h (hunk 7 above)
Comment (apivovarov) [reply include/dlr_common.h]: fixed

File: include/dlr_common.h (hunk 8 above)
Comment (apivovarov) [reply include/dlr_common.h]: fixed

File: include/dlr_tvm.h [hunk 9]
```diff
@@ -0,0 +1,45 @@
+#ifndef DLR_TVM_H_
+#define DLR_TVM_H_
+
+#include <graph/graph_runtime.h>
+#include "dlr_common.h"
+
+namespace dlr {
+
+/*! /brief Get the paths of the TVM model files.
```
Comment (zhiics) [inline include/dlr_tvm.h]: shouldn't be `\brief`? plz fix the other places as well?

Comment (apivovarov) [reply include/dlr_tvm.h]: fixed everywhere

**Pull request merged**